### ISSUE TITLE
`archive-post` hook runs before archive?

### ISSUE THREAD

**Issue by dlford** [offen/docker-volume-backup#187]
**Describe the bug**
I would expect `archive-post` to run after archiving, but it seems to run before causing backup to fail.

**To Reproduce**
```yml
version: "2.2"
services:
  # ...
  kibana_rsync:
    container_name: elk_kibana_rsync
    tty: true
    image: eeacms/rsync
    restart: unless-stopped
    labels:
      - docker-volume-backup.exec-label=elk
      - docker-volume-backup.archive-pre=sh -c "rsync -aAX --ignore-missing-args --delete-missing-args /data/ /bu/"
      - docker-volume-backup.archive-post=sh -c "rm -rf /bu/*"
    volumes:
      - kibana:/data:ro
      - kibana_bu:/bu
  backup:
    container_name: elk_backup
    image: offen/docker-volume-backup:latest
    restart: unless-stopped
    environment:
      BACKUP_FILENAME: backup.tar.gz
      NOTIFICATION_URLS: "..."
      NOTIFICATION_LEVEL: error
      EXEC_LABEL: elk
      AWS_ENDPOINT: "..."
      AWS_S3_BUCKET_NAME: "..."
      AWS_ACCESS_KEY_ID: "..."
      AWS_SECRET_ACCESS_KEY: "..."
    volumes:
      - kibana_bu:/backup/kibana:ro
      - /var/run/docker.sock:/var/run/docker.sock
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
volumes:
  kibana:
  kibana_bu:
```

**Expected behavior**
`archive-post` should run after tar archive is created.

**Version (please complete the following information):**
 - Image Version: latest (`sha256:837cbec53ba8fcaec3ad1f9d2a25086ad33fdbfa90585dc9f92e1947e1dea15d`)
 - Docker Version: 23.0.0
 - Docker Compose Version (if applicable): 23.0.0

**Additional context**

Here are the logs for the most recent run showing the `archive-pre` command executing, followed by `archive-post`, and then the archive step fails because the files are all gone.

```
time="2023-02-10T00:00:32-07:00" level=info msg="Running docker-volume-backup.archive-pre command sh -c \"rsync -aAX --ignore-missing-args --delete-missing-args /data/ /bu/\" for container elk_kibana_rsync"
time="2023-02-10T00:11:12-07:00" level=info msg="Running docker-volume-backup.archive-post command sh -c \"rm -rf /bu/*\" for container elk_kibana_rsync"
time="2023-02-10T00:11:12-07:00" level=error msg="Fatal error running backup: createArchive: error compressing backup folder: createArchive: error creating archive: compress: error writing /backup/kibana/chromium-26SeJH to archive: writeTarGz: error getting file infor for /backup/kibana/chromium-26SeJH: lstat /backup/kibana/chromium-26SeJH: no such file or directory"
time="2023-02-10T00:11:12-07:00" level=info msg="Removed tar file `/tmp/backup.tar.gz`."crond: crond (busybox 1.35.0) started, log level 8
````

**Comment by m90** [comment]
What you are seeing is rather confusing, but is the expected behavior and also normal: `post` commands always run, even if the wrapped lifecycle phase fails (which is what is happening in your case): https://github.com/offen/docker-volume-backup/blob/2d37e08743c1547f79fdda30e084501bf0c5e9ee/cmd/backup/exec.go#L190-L203

Due to some Go internals I was too lazy to work around yet, this scenario also causes the log lines to appear in the order you are observing in your setup:

- pre commands run successfully and log
- the archive operation fails, error message in your case is  `compress: error writing /backup/kibana/chromium-26SeJH to archive: writeTarGz: error getting file infor for /backup/kibana/chromium-26SeJH: lstat /backup/kibana/chromium-26SeJH: no such file or directory"`. This error isn't logged but passed further up instead
- the post command is run successfully and logs
- the top level error handler receives the error your failed archive operation created and logs that error

I.e.: even if the log output doesn't look like it, the commands are still run in the correct order, the reason that makes your backup run fail is not related to the command mechanism, but something else.

As for why writing your archive fails, I am not sure right now, it's probably that the file system you are trying to back up is volatile at the time of archiving.

---

On a side note: why do you need to use the `rsync` container for creating a snapshot of the data you would like to backup? You could probably do the same using `cp`, which would allow you to do everything at once in the `backup` container itself. This is described here: https://github.com/offen/docker-volume-backup#replace-deprecated-backup_from_snapshot-usage

**Comment by dlford** [comment]
@m90 Thanks for the explanation, that makes sense.

I'm at a loss now on how to back this container up without stopping it though, I tried `cp` first and was getting an error from the `cp` command that files weren't found, so I went on to `rsync` and it still fails.

But here's the thing, there is no reason for that filesystem to be volatile during a backup, the `kibana_bu` volume is only mounted to the `rsync` container and the backup container. Is it possible that `docker-volume-backup` starts archiving before `rsync` finishes running? I don't see any other explanation for this behavior?

**Comment by m90** [comment]
> Is it possible that docker-volume-backup starts archiving before rsync finishes running? 

Yeah, it is pretty mysterious as the filesystem really should be stable when using rsync, but I also don't see any reason why `cp` would fail. I think the error you see when trying to use `cp` is probably caused by the same underlying issue as when you are using `rsync`.

> I tried cp first and was getting an error from the cp command that files weren't found

Could you post the exact error message you are seeing here? Maybe this helps us with finding the root cause of all this.

**Comment by dlford** [comment]
@m90 I get the same error with `cp`, but with that addition of many like this from the `cp` command itself:

```
cp: can't stat '/data/chromium-23SaJQ': No such file or directory
```

Side note, I wonder if it's permissions related? The images for Elasticsearch and Kibana run as user `1000` instead of root, I don't think that's it, but it might be worth exploring if there's nothing else obvious to track down.

**Comment by m90** [comment]
I guess there is something very special about the directory you want to backup in as there are files that get picked up by `cp -R`, but then you can't call `stat` on them. No idea what might be causing this. I doubt it's permissions as `root` should be able to copy everything, also files owned by `1000`.

Could you inspect such a file (e.g. `/data/chromium-23SaJQ`) further and see if there is anything special about it?

**Comment by dlford** [comment]
@m90 I don't see anything that looks off here:

```
root:~# mkdir bu
root:~# docker run --rm -v /root/bu:/bu -v elk_kibana:/data:ro eeacms/rsync sh -c "rsync -aAX --ignore-missing-args --delete-missing-args /data/ /bu/"
root:~# stat bu/chromium-26SeJH
  File: bu/chromium-26SeJH
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: fd00h/64768d    Inode: 656942      Links: 2
Access: (0700/drwx------)  Uid: ( 1000/     admin)   Gid: ( 1000/     admin)
Access: 2023-02-11 16:40:01.287120055 -0700
Modify: 2023-01-27 14:14:55.473619621 -0700
Change: 2023-02-11 16:40:01.287120055 -0700
 Birth: 2023-02-11 16:40:00.523085593 -0700
```

Something else interesting, it seems to back up fine if I manually run it, but every night it fails with the same error when running on the schedule:

```
root:~# docker exec elk_backup backup
time="2023-02-11T16:44:13-07:00" level=info msg="Running docker-volume-backup.archive-pre command sh -c \"rsync -aAX --ignore-missing-args --delete-missing-args /data/ /bu/\" for container elk_kibana_rsync"
time="2023-02-11T16:45:57-07:00" level=info msg="Created backup of `/backup` at `/tmp/backup.tar.gz`."
time="2023-02-11T16:45:57-07:00" level=info msg="Running docker-volume-backup.archive-post command sh -c \"rm -rf /bu/*\" for container elk_kibana_rsync"
time="2023-02-11T16:46:12-07:00" level=info msg="Removed tar file `/tmp/backup.tar.gz`."
time="2023-02-11T16:46:12-07:00" level=info msg="Finished running backup tasks."
```

**Comment by m90** [comment]
> So it only fails when running on the schedule, and it's been failing every time.

Is there anything running in parallel when this happens? The command that is run by `crond` is the exact same as when executing it manually, so it must me some external factor that is causing this, something that was not present when you ran the command manually.

It's hard for me to give further advice on how to debug this right now. I'd probably exclude `rsync` from the equation for now and just use `cp`  as it seems that `cp` and archive creation fail because of the same condition, using `rsync` will just make the setup fail later.

**Comment by dlford** [comment]
@m90 If I remove `rsync`, the backups will always fail because the filesystem to be backed up is volatile and `cp` will return exit code `1` because it couldn't find one or more of the files (`cp` gets a list of files to copy and then starts copying). `docker-volume-backup` will bail out if that return code isn't `0`. Using `rsync` with `--ignore-missing-args --delete-missing-args` will suppress the return code `24` for missing files and return with `0` so the backup process can continue.

I'll try changing the schedule and see if that makes a difference.

Thanks for your help looking into this!

**Comment by m90** [comment]
> cp gets a list of files to copy and then starts copying

Do you use `cp` with a glob pattern, or do you `cp -R path/to/dir`?

**Comment by dlford** [comment]
I do `cp -R`, but when the files change it logs `cp: can't stat '/data/chromium-23SaJQ': No such file or directory`, then the exit code is not `0`, and the backup aborts.

You might be right about the time, I scheduled it for 10am this morning and it seems to have run without issue. I wonder if the other instances of `docker-volume-backup` are interfering somehow? But it only seems to affect the one instance, all the others run fine at midnight. Either way I have a workaround, thanks again!

**Comment by m90** [comment]
Closing this as you seem to have gotten it working? If this resurfaces somehow, feel free to reopen.

**Comment by dlford** [comment]
@m90 I have an update on this, and I think I've found the problem.

Initially, I had only converted a few backups to use the pre / post hooks, but after the schedule change workaround seemed stable I started updating a few more backups to use the hooks, and the different time schedule.

I then started getting multiple emails at midnight for failed backups, all the emails listed all the pre / post hooks for all the backups. This was very confusing because all the backups with hooks run before midnight.

It finally dawned on me just now that the backups that aren't set up for hooks and are still running at midnight are trying to run all the hooks for all the backups, because they are missing the `EXEC_LABEL` environment variable, which may be the cause of my initial troubles. Sure enough, when I logged in to Backblaze there were several backups missing, but all of the backups with hooks set up were there, it was the ones that haven't been updated for hooks that were missing.

I've added the `EXEC_LABEL` variable to all backup containers with a unique value, I'm pretty confident that this will resolve the issue, I will know for sure tonight when the backups run again.
